In [1]:
import tensorflow as tf
print(tf.__version__)

2.17.0


In [3]:
import os
print(os.getcwd())

C:\Users\Admin


In [5]:
df = pd.read_csv(r"C:\Users\Admin\Downloads\Alphabets_data.csv")

In [6]:
print(df.head())
print(df.shape)

  letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10  
(20000, 17)


In [7]:
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler

X = df.iloc[:, 1:]
y = df.iloc[:, 0]

le = LabelEncoder()
y = le.fit_transform(y)

scaler = StandardScaler()
X = scaler.fit_transform(X)

In [8]:
import pandas as pd

In [10]:
import pandas as pd

df = pd.read_csv(r"C:\Users\Admin\Downloads\Alphabets_data.csv")

print(df.head())
print(df.shape)

  letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10  
(20000, 17)


In [11]:
pd.read_csv(r"C:\Users\Admin\Downloads\Alphabets_data.csv")

,letter,xbox,ybox,width,height,onpix,xbar,ybar,x2bar,y2bar,xybar,x2ybar,xy2bar,xedge,xedgey,yedge,yedgex
0,T,2,8,3,5,1,8,13,0,6,6,10,8,0,8,0,8
1,I,5,12,3,7,2,10,5,5,4,13,3,9,2,8,4,10
2,D,4,11,6,8,6,10,6,2,6,10,3,7,3,7,3,9
3,N,7,11,6,6,3,5,9,4,6,4,4,10,6,10,2,8
4,G,2,1,3,1,1,8,6,6,6,6,5,9,1,7,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19995,D,2,2,3,3,2,7,7,7,6,6,6,4,2,8,3,7
19996,C,7,10,8,8,4,4,8,6,9,12,9,13,2,9,3,7
19997,T,6,9,6,7,5,6,11,3,7,11,9,5,2,12,2,4
19998,S,2,3,4,2,1,8,7,2,6,10,6,8,1,9,5,8


In [12]:
## Train-Test Split
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print(X_train.shape, X_test.shape)

(16000, 16) (4000, 16)


In [13]:
## Build Neural Network (TensorFlow)
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model = Sequential()

model.add(Dense(64, activation='relu', input_shape=(X.shape[1],)))
model.add(Dense(32, activation='relu'))
model.add(Dense(26, activation='softmax'))  # 26 letters

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

model.summary()

C:\Users\Admin\anaconda3\envs\ann_env\lib\site-packages\keras\src\layers\core\dense.py:95: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 64)                  │           1,088 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 32)                  │           2,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 26)                  │             858 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 4,026 (15.73 KB)

 Trainable params: 4,026 (15.73 KB)

 Non-trainable params: 0 (0.00 B)

In [14]:
## Train Model
history = model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_data=(X_test, y_test)
)

Epoch 1/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.4901 - loss: 1.8761 - val_accuracy: 0.7065 - val_loss: 1.0606
Epoch 2/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7459 - loss: 0.8922 - val_accuracy: 0.7837 - val_loss: 0.7572
Epoch 3/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7958 - loss: 0.6939 - val_accuracy: 0.8188 - val_loss: 0.6281
Epoch 4/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8272 - loss: 0.5858 - val_accuracy: 0.8465 - val_loss: 0.5436
Epoch 5/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8486 - loss: 0.5095 - val_accuracy: 0.8597 - val_loss: 0.4872
Epoch 6/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8626 - loss: 0.4518 - val_accuracy: 0.8712 - val_loss: 0.4405
Epoch 7/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8805 - loss: 0.4068 - val_accuracy: 0.8742 - val_loss: 0.4092
Epoch 8/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8874 - loss: 0.3709 - val_accuracy: 0.

In [15]:
## Evaluate Model
loss, accuracy = model.evaluate(X_test, y_test)
print("Accuracy:", accuracy)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.9055 - loss: 0.3190
Accuracy: 0.9054999947547913


In [16]:
## TRAIN THE MODEL
history = model.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_data=(X_test, y_test)
)

Epoch 1/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9104 - loss: 0.2958 - val_accuracy: 0.9045 - val_loss: 0.3157
Epoch 2/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9162 - loss: 0.2771 - val_accuracy: 0.9155 - val_loss: 0.2842
Epoch 3/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9216 - loss: 0.2603 - val_accuracy: 0.9160 - val_loss: 0.2799
Epoch 4/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9261 - loss: 0.2464 - val_accuracy: 0.9130 - val_loss: 0.2705
Epoch 5/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9287 - loss: 0.2343 - val_accuracy: 0.9222 - val_loss: 0.2583
Epoch 6/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9333 - loss: 0.2226 - val_accuracy: 0.9197 - val_loss: 0.2598
Epoch 7/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9351 - loss: 0.2131 - val_accuracy: 0.9268 - val_loss: 0.2432
Epoch 8/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9362 - loss: 0.2057 - val_accuracy: 0.

In [17]:
## Try Model 2 (change neurons)
model2 = Sequential()

model2.add(Dense(128, activation='relu', input_shape=(X.shape[1],)))
model2.add(Dense(64, activation='relu'))
model2.add(Dense(26, activation='softmax'))

model2.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

history2 = model2.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_data=(X_test, y_test)
)

loss2, acc2 = model2.evaluate(X_test, y_test)
print("Model 2 Accuracy:", acc2)

Epoch 1/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.6226 - loss: 1.3998 - val_accuracy: 0.7905 - val_loss: 0.7856
Epoch 2/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8064 - loss: 0.6716 - val_accuracy: 0.8400 - val_loss: 0.5690
Epoch 3/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8539 - loss: 0.5029 - val_accuracy: 0.8767 - val_loss: 0.4398
Epoch 4/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8826 - loss: 0.4021 - val_accuracy: 0.8950 - val_loss: 0.3654
Epoch 5/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9018 - loss: 0.3369 - val_accuracy: 0.9100 - val_loss: 0.3206
Epoch 6/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9139 - loss: 0.2885 - val_accuracy: 0.9175 - val_loss: 0.2849
Epoch 7/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9268 - loss: 0.2518 - val_accuracy: 0.9252 - val_loss: 0.2583
Epoch 8/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9346 - loss: 0.2207 - val_accuracy: 0.

In [18]:
## Try Model 3 (change optimizer)
model3 = Sequential()

model3.add(Dense(64, activation='relu', input_shape=(X.shape[1],)))
model3.add(Dense(32, activation='relu'))
model3.add(Dense(26, activation='softmax'))

model3.compile(
    optimizer='sgd',   # changed
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

history3 = model3.fit(
    X_train, y_train,
    epochs=10,
    batch_size=32,
    validation_data=(X_test, y_test)
)

loss3, acc3 = model3.evaluate(X_test, y_test)
print("Model 3 Accuracy:", acc3)

Epoch 1/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.1415 - loss: 3.0144 - val_accuracy: 0.2648 - val_loss: 2.7026
Epoch 2/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.3517 - loss: 2.3997 - val_accuracy: 0.4460 - val_loss: 2.0877
Epoch 3/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5039 - loss: 1.8527 - val_accuracy: 0.5530 - val_loss: 1.6438
Epoch 4/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.5868 - loss: 1.5004 - val_accuracy: 0.6252 - val_loss: 1.3773
Epoch 5/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6408 - loss: 1.2928 - val_accuracy: 0.6680 - val_loss: 1.2123
Epoch 6/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.6782 - loss: 1.1601 - val_accuracy: 0.6975 - val_loss: 1.1035
Epoch 7/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7034 - loss: 1.0652 - val_accuracy: 0.7220 - val_loss: 1.0171
Epoch 8/10
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7243 - loss: 0.9883 - val_accuracy: 0.

In [19]:
## Try Model 4 (increase epochs)
history4 = model.fit(
    X_train, y_train,
    epochs=20,
    batch_size=32,
    validation_data=(X_test, y_test)
)

Epoch 1/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9439 - loss: 0.1818 - val_accuracy: 0.9300 - val_loss: 0.2226
Epoch 2/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9453 - loss: 0.1743 - val_accuracy: 0.9350 - val_loss: 0.2161
Epoch 3/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9465 - loss: 0.1691 - val_accuracy: 0.9373 - val_loss: 0.2099
Epoch 4/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9479 - loss: 0.1632 - val_accuracy: 0.9373 - val_loss: 0.2087
Epoch 5/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9494 - loss: 0.1604 - val_accuracy: 0.9392 - val_loss: 0.2033
Epoch 6/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9498 - loss: 0.1562 - val_accuracy: 0.9408 - val_loss: 0.1941
Epoch 7/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9526 - loss: 0.1489 - val_accuracy: 0.9385 - val_loss: 0.1990
Epoch 8/20
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9544 - loss: 0.1446 - val_accuracy: 0.

| Model   | Changes       | Accuracy |
| ------- | ------------- | -------- |
| Model 1 | Default       | 90.5%    |
| Model 2 | More neurons  | ?        |
| Model 3 | SGD optimizer | ?        |
| Model 4 | More epochs   | ?        |



In [20]:
from sklearn.metrics import classification_report

y_pred = model.predict(X_test)
y_pred = y_pred.argmax(axis=1)

print(classification_report(y_test, y_pred))

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       149
           1       0.86      0.93      0.89       153
           2       0.95      0.94      0.95       137
           3       0.94      0.97      0.95       156
           4       0.90      0.95      0.92       141
           5       0.98      0.91      0.94       140
           6       0.96      0.93      0.94       160
           7       0.93      0.77      0.84       144
           8       0.94      0.96      0.95       146
           9       0.97      0.94      0.96       149
          10       0.94      0.89      0.91       130
          11       0.98      0.95      0.97       155
          12       0.98      0.95      0.97       168
          13       0.98      0.93      0.95       151
          14       0.92      0.97      0.94       145
          15       0.98      0.95      0.96       173
          16       0.98      0.98      0

## Conclusion 
Model with Adam performed better
Increasing neurons improved accuracy
Too many epochs → overfitting risk
Best accuracy achieved = ___%